File: palindromic-substrings/palindromic-substrings.java
class Solution {
    public int countSubstrings(String s) {
        int n = s.length();
        boolean[][] dp = new boolean[n][n];
        int count = 0;
        for (int i = n - 1; i >= 0; i--) {
            char a = s.charAt(i);
            for (int j = i; j < n; j++) {
                char b = s.charAt(j);
                
                if (i == j) {
                    dp[i][j] = true;
                } else if (a == b) {
                    if (j == i + 1) {
                        dp[i][j] = true;
                    } else if (dp[i + 1][j - 1]) {
                        dp[i][j] = true;
                    }
                }
                
                if (dp[i][j]) {
                    count += 1;
                }
            }
        }
        
        return count;
    }
}

File: palindromic-substrings/palindromic-substrings.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        
        count = 0
        def helper(i):
            nonlocal count
            left = i
            right = i
            
            while left >= 0 and right < len(s):
                if s[left] == s[right]:
                    count += 1
                    left -=1
                    right += 1
                else:
                    break
            
            left = i
            right = i + 1
            while left >= 0 and right < len(s):
                if s[left] == s[right]:
                    count += 1
                    left -=1
                    right += 1
                else:
                    break
        
        for i in range(len(s)):
            helper(i)
        
        return count
            